Enhancement: Adding Link State Management for vNICs

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

orelmisan commented:
/cc @EdDev @nirdothan @ormergi

ormergi commented on sig-network/link-state-proposal.md:
````diff
@@ -0,0 +1,191 @@
+The following values will be added to specify the required link state:
+```go
+const (
+    InterfaceStateLinkUp   InterfaceState = "linkUp"
+    InterfaceStateLinkDown InterfaceState = "linkDown"
````
I would avoid having "link" in the naming at this abstraction layer, having the naming more generic doesnt make me think whether there are some other options affecting other stuff aspects of the interface (e.g.: network-manager device and link, disabling the link does mean device is disabled).

Same goes for the new attribute in the status, calling it just "state".

orelmisan replied:
The existing enumeration is `absent`, which relates to the hot-unplug functionality.
By having the `link` prefix, we relate the new enumerations to the proposed functionality (setting the link state).

ormergi replied:
It can be done without the link in the naming, with simple names.
I dont think it contradicts with "absent" semantics.

orelmisan replied:
Changed the possible values to `up` and `down`.

ormergi commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,191 @@
+   - When VM is running
+   - On hotplug
+4. Addition of e2e tests.
+5. Addition of the `LinkState` field.
```
Does it refer to the status?

orelmisan replied:
Yes.

ormergi reviewed (commented):
I had one comment on the attributes naming, other than that LGTM.

ormergi commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,191 @@
+# Overview
+
+The purpose of this design proposal is to introduce support for link state management (up/down) in KubeVirt. This feature
+will enable VM owners to dynamically control the link state. It aligns KubeVirt with traditional virtualization platforms
```
I think they meant for virtualization plaftom that is not container based (such as Kubevirt).

lyarwood reviewed (commented):
Many thanks for putting this excellent proposal forward as something we can use to try out the new process in v1.5.0 @orelmisan. I do wonder however if the original design proposals PR should still be used to discuss the actual design itself to keep discussions in here focused on the new process and structure of the enhancements proposals.

Fully appreciate that's doubling your work in some ways to keep the rough design in sync between the two repos but I'm just worried about process discussions around ehancements being lost among valid technical discussions on the design.

Feel free to ignore if others are happy handling both here.

orelmisan commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,191 @@
+and enhances its utility in environments requiring precise network state control.
+
+By introducing link state management, KubeVirt will:
+- Provide VM-level controls to configure the link state of individual interfaces.
```
Users are expected to interact mostly with the VirtualMachine CR, while the VirtualMachineInstance CR is an implementation detail.
Having said that, there are some status reports which are only visible on the VMI.

orelmisan commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,191 @@
+
+## Non Goals
+
+1. Affect network plumbing on the host - the setup process will remain unchanged change.
```
Yes, thanks.

orelmisan commented on sig-network/link-state-proposal.md:
````diff
@@ -0,0 +1,191 @@
+}
+```
+
+Type `InterfaceState` is an enum, currently having a single value:  
````
There are no emums in Go, but I think the intention is clear.
Feel free to suggest an alternative wording.

orelmisan replied:
Done.

orelmisan commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,191 @@
+An empty value will be considered as `linkUp`.
+
+The network validator will be adjusted to allow the two new values for:
+1. All interfaces connected to primary and secondary networks.
+2. All bindings - core and plugins.
```
> Doesn't this basically mean all interfaces?

It does, but I wanted to emphasize the fact that this functionality could work on interfaces connected to primary and secondary networks (even on default Multus networks).

Feel free to suggest an alternative wording.

orelmisan replied:
Done.

orelmisan commented:
> Many thanks for putting this excellent proposal forward as something we can use to try out the new process in v1.5.0 @orelmisan. I do wonder however if the original design proposals PR should still be used to discuss the actual design itself to keep discussions in here focused on the new process and structure of the enhancements proposals.
> 
> Fully appreciate that's doubling your work in some ways to keep the rough design in sync between the two repos but I'm just worried about process discussions around ehancements being lost among valid technical discussions on the design.
> 
> Feel free to ignore if others are happy handling both here.

Thank you for your kind words @lyarwood! much appreciated.

I suggest having both discussions here in order to not double the author and reviewers' efforts.

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/7a0656d3609574b03f9a3bfc03c707bb1b6d77e7..ca0a240c69ead7733f813bac3ad7301744c7ce60): Addressed @ormergi's [comment](https://github.com/kubevirt/enhancements/pull/2#discussion_r1908653207) and @nirdothan's [comment](https://github.com/kubevirt/enhancements/pull/2#discussion_r1908653128).

lyarwood commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,191 @@
+status:
+  interfaces:
+    - name: default
+      linkState: down
```
Requesting `linkDown` but getting `down` is also confusing IMHO.

orelmisan replied:
Adjusted the allowed values to `up` and `down`.

lyarwood reviewed (commented):
> Thank you for your kind words @lyarwood! much appreciated.
> 
> I suggest having both discussions here in order to not double the author and reviewers' efforts.

That's fine but there's no requirement for this to land in v1.5.0 and it might easily become blocked by process discussions, however if you're okay with that we can proceed.

I've added some comments and asked if you could align with the new proposal and issue templates we landed yesterday. Feedback, PRs, issues etc welcome on both of these obviously from anyone looking at this PR.

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/ca0a240c69ead7733f813bac3ad7301744c7ce60..9a480e60c3c48358efc2d543b90dbd5c0d37d9d8): Adjusted the document's structure per @lyarwood's [comment](https://github.com/kubevirt/enhancements/pull/2#discussion_r1910066938).

orelmisan commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,191 @@
+
+# Overview
+
+The purpose of this design proposal is to introduce support for link state management (up/down) in KubeVirt. This feature
```
Done.

orelmisan commented on sig-network/link-state-proposal.md:
````diff
@@ -0,0 +1,191 @@
+```go
+type VirtualMachineInstanceNetworkInterface struct {
+	...
+	LinkState string `json:"linkState,omitempty"`
````
I've dropped the `link` prefix in order to not add another field.

orelmisan commented:
[Change 1](https://github.com/kubevirt/enhancements/compare/9a480e60c3c48358efc2d543b90dbd5c0d37d9d8..1d531808c6f484401042e839f520431cf97178ff): Addressed @nirdothan's and @lyarwood's comments.

[Change 2](https://github.com/kubevirt/enhancements/compare/1d531808c6f484401042e839f520431cf97178ff..3327ac32dd147eefe6c3efd59bc3e7e7463b58e1): Fixed a typo.

yossisegev commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,197 @@
+An empty value will be considered as `up`.
+
+The network validator will be adjusted to allow the two new values for all interfaces regardless of whether they are 
+connected to primary or secondary networks, or whether they use a core binding or a binding plugin.
```
1. Setting A VM's masquerade (pod) interface to `down` - is it even legit? How will it actually affect the guest-to-pod connectivity?
2. In UDN VM - will setting the primary interface to `down` only affect the node-facing part of the interface, and keep the pod networking fully functioning?

orelmisan replied:
Which according to [1]:
> If 'down' is specified as the value, the interface behaves as if it had the network cable disconnected.

[1] https://libvirt.org/formatdomain.html#modifying-virtual-link-state

yossisegev commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,197 @@
+## Feature lifecycle Phases
+
+### GA
+Since this feature is small and has no significant risk, it could be released directly as GA.
```
Let's test it a little before determining that... :) 
An innocent-looking feature might turn out to be a hidden monster...
And specifically with this feature - the ability to set the guest's pod networking down worries me.

orelmisan replied:
> Let's test it a little before determining that... :)

E2E tests are planned, and the test case / scenarios are written down.
Do you think additional scenarios should be added?

orelmisan replied:
> And specifically with this feature - the ability to set the guest's pod networking down worries me.

Could you please elaborate why does it worries you?
The user has full control over this field, and the system has none - so it shouldn't automatically change.

orelmisan commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,191 @@
+Adding Link State Management for vNICs
```
Done.

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/3327ac32dd147eefe6c3efd59bc3e7e7463b58e1..ca55afeeef34951bb509a8f981ffa9f3892f77c1): Added a note about link state status report.

EdDev commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,201 @@
+
+## User Stories
+
+As a VM owner I would like to be able to:
```
The pod network is special, can we get confirmation from the stakeholders if it makes sense to have the primary network be toggled like this? I mainly want to make sure there is no special expectations about this.
Disconnecting from the pod network, which acts as a control plane as well, is not that trivial IMO.

orelmisan replied:
@phoracek what do you think?

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/ca55afeeef34951bb509a8f981ffa9f3892f77c1..f990d8da38f11cfc909c938d7bc7012d738671f8): Added an exception for SR-IOV NICs - as we cannot control their link state via the domain.

vladikr commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,211 @@
+> [!NOTE]
+> For additional details please see libvirt's [documentation](https://libvirt.org/formatdomain.html#modifying-virtual-link-state).
+
+virt-launcher's `LibvirtDomainManager.SyncVMI` will be extended to support updating the interfaces link state while the VM is running.
```
It's not entirely clear to me how this process will work. Could you please elaborate and explain the flow?

lyarwood commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,211 @@
+## Release Signoff Checklist
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)
```
^ As above, I'll try to make this more clear now.

orelmisan replied:
I've added an `X` inside the checkbox.
Not sure what else has to be done.

lyarwood commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,211 @@
+6. Adjusting virt-handler's reporting logic.
+7. Adjusting e2e tests.
+
+## Feature lifecycle Phases
```
The feature lifecycle document [1] suggests that we always start with a non-GA FG and then graduate. 

While I can see an argument to skip this requirement entirely for this feature and just GA it straight away I wonder if we should instead compromise with a beta FG for the v1.5.0 release we can then GA in v1.6.0?

@EdDev @vladikr @xpivarc I'd appreciate your thoughts here.

IMHO we should also extend the template and/or lifecycle document [1] with the possibility of starting with a beta FG if folks agree to it here.

[1] https://github.com/kubevirt/community/blob/main/design-proposals/feature-lifecycle.md

EdDev replied:
I am in favor to be strict with regards of new features, however, in case a proper reasoning can be provided, we can enforce an exception on the FG path.
I think this one is a fit to bypass any FG, it is too trivial to require an evaluation/experiment stage.

I asked to give a more detailed convincing explanation on why GA first is a good fit [1].

I vote to directly approve this small feature for GA.

[1] https://github.com/kubevirt/enhancements/pull/2#discussion_r1918986205

orelmisan replied:
I've added a few points, following this comment and @EdDev's [comment](https://github.com/kubevirt/enhancements/pull/2#discussion_r1918986205).

lyarwood commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,211 @@
+
+## Feature lifecycle Phases
+
+### GA
```
Can you also include the proposed version for the given milestone?

`### GA (v1.5.0)` etc.

orelmisan replied:
Done.

orelmisan commented on sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,211 @@
+# VEP #2: Adding Link State Management for vNICs
```
Done.

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/1137505c78b69498c87292630a8f87bf9192de51..fc47890ce3ea8005bba78dc67628c48a2d6ae303): Addressed @vladikr's and @lyarwood's comments.

orelmisan commented:
/sig network

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/fc47890ce3ea8005bba78dc67628c48a2d6ae303..349f2edd5aae39cd4962d904b6aa43d0e3fd8f74): Rebase + moved the file under the `veps` directory.

phoracek commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,217 @@
+- Start the VM with one or more interfaces with their link set to `down`.
+- Toggle the link state of one or more interfaces of a running VM.
+- Hot plug an interface with a link state set to `down`.
+- Control the link state of interfaces connected to primary and secondary networks.
```
To follow-up on our offline discussion - I do see a value in having it on the primary network from the start. The motivation to isolate a running VM from the rest of the network applies for both primary and secondary. If there are any strong specific concerns about link-state control causing issues to primary CNIs, please raise them @maiqueb @EdDev

EdDev replied:
On the primary network, we heavily depend on dynamic IP configuration served by Kubevirt (in the virt-launcher).
Moreover, IPv6 full support requires cloud-init intervention.

Unlike the secondary networks which we depend much more on the VM owner and operator, on the primary we have more responsibility.
As I see it, that increases the risk and with it the need to cover much more scenarios than on secondary networks.
Said that, based on preliminary checks, we can decide to take the risk and warn about the implications of using it on the primary network without blocking it.

phoracek replied:
Could you share more about those two examples? I don't understand how IP configuration serving or IPv6 is affected by the link state.

EdDev replied:
For dynamic configuration (e.g. DHCP) the client needs to successfully re-request to be assign an IP.
The server side (Kubevirt) needs to successfully answer it.
Or some network provider thing (OVN-K?) needs to answer it.

Same goes with configuration coming from cloud-init (e.g. for IPv6), it needs to be removable and it presents different scenarios: When link was down on boot and came up later, when link went down and up while VM is running, etc...

phoracek replied:
How are those scenarios different from soft reboot, which we do support today with all network options?

EdDev replied:
Please resolve this thread.

EdDev commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,217 @@
+## Feature lifecycle Phases
+
+### GA (v1.5.0)
+Since this feature is small and has no significant risk, it could be released directly as GA.
```
There is a need for a a bit more information in order to be convincing.

Reporting the link state through the interface status is a natural addition to the interface details.
Today, the vNIC link is always `up` and even without changes to the desired configuration, it can always be reported as such and be correct.
The option of reporting the link as `down` is added toghether with the extension of the spec with the new values or an existing field.
Finally, the input link state is using an existing field. The new values and implementation is straight-forward with no special implication that may be considered as a risk. Both from the API changes and functionality.

orelmisan replied:
Done.

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/349f2edd5aae39cd4962d904b6aa43d0e3fd8f74..afa800ce1565b516125a645ca42eca195dc9128e): Addressed @EdDev's comments.

orelmisan commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,217 @@
+
+## Overview
+
+The purpose of this enhancement is to introduce support for link state management (up/down) in KubeVirt. This feature
```
Changed the wording to give a better context.

orelmisan commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,217 @@
+connected to primary or secondary networks, or whether they use a core binding or a binding plugin.
+
+> [!IMPORTANT]
+> Controlling the link state of SR-IOV NICs will not be supported as we cannot control their link state via the domain.
```
Good idea!
Done.

orelmisan commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,217 @@
+An empty specification (as exists today) would be considered as `up`.
+
+> [!NOTE]
+> The linkState information will be reported for interfaces which are present in the VirtualMachineInstance's spec.
```
There is no problem reporting the link state of SR-IOV NICS.
They will just be always `up`.
Added an explanation to clarify.

EdDev replied:
I think this is incorrect.
Unlike the "regular" vNIC, SR-IOV VFs are dependent on the PF HW configuration.

From `man ip-link`:

>                       state auto|enable|disable - set the virtual link
>                       state as seen by the specified VF. Setting to auto
>                       means a reflection of the PF link state, enable
>                       lets the VF to communicate with other VFs on this
>                       host even if the PF link state is down, disable
>                       causes the HW to drop any packets sent by the VF.

Bottom line, the link is not assured to be operational `up`, it depends on how it is configured through the SR-IOV CNI and in the `auto` case, dependent on the actual PF operational link state.

orelmisan replied:
Thank you, I wasn't aware to this fact.
Done.

EdDev replied:
Please resolve

orelmisan commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,217 @@
+> The linkState information will be reported for interfaces which are present in the VirtualMachineInstance's spec.
+> Interfaces defined inside the guest will not have their link state reported.
+>
+> Changing the link state of an interface inside the guest or in the underlying network provider - will not affect this report.
```
Done.

EdDev reviewed (commented):
Thanks!
Except the SR-IOV link state reporting, all seem good to me.

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/afa800ce1565b516125a645ca42eca195dc9128e..ef9e6bbb63a006698557f64bb5750aba1cedc9f5): SR-IOV NICs' link state will not be reported.

EdDev reviewed (approved):
Thank you!

Approving for sig-network.
/approve

maiqueb commented:
/hold

Please allow me a few days to review this PR - at least the part about the UDN interaction. 

I'm a bit swamped at the moment, thus I cannot promise when; I apologize in advance .

jean-edouard commented on veps/sig-network/link-state-proposal.md:
````diff
@@ -0,0 +1,232 @@
+      linkState: down
+```
+
+## Alternatives
````
Nit: I think it's better to remove empty sections. (or at least add a line explaining why it's irrelevant like below)

orelmisan replied:
It was removed.

jean-edouard commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,232 @@
+
+> [!NOTE]
+> The linkState information will be reported for interfaces which are present in the VirtualMachineInstance's spec.
+> Interfaces defined inside the guest will not have their link state reported.
```
What about the default interface when none is defined in the spec, will its state be reported?

orelmisan replied:
The default interface is automatically added to the VMI spec (unless the use had explicitly specified against it).

https://github.com/kubevirt/kubevirt/blob/a7c2595a54149662e564d89fd377f60b210570a2/pkg/network/vmispec/defaults.go#L33

orelmisan replied:
So it will be reported.

jean-edouard reviewed (approved):
I think this is a great idea.
Just a couple, very minor, questions in-line.

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/ef9e6bbb63a006698557f64bb5750aba1cedc9f5..39c035213a77fc3e34333de1ae0f76360c071365): Removed an empty section.

vladikr commented:
Thanks, everyone for signing off on this.
I'm happy to approve this.
/approve

@maiqueb feel free to remove the hold once you are ready.
Thanks!

xpivarc commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,230 @@
+virtualization platforms and enhances its utility in environments requiring precise network state control.
+
+By introducing link state management, KubeVirt will:
+- Provide VM-level controls to configure the link state of individual interfaces.
```
Please be specific on the resource, this sounds like VM specific field rather than VMI.

orelmisan replied:
This field appears both in the VM and the VMI.
A user can only edit it on the VM.

xpivarc replied:
Let's make it more clear please.

orelmisan replied:
How would you phrase it?

xpivarc replied:
Maybe mention this in non-goals?

xpivarc commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,230 @@
+## Non Goals
+
+1. Affect network plumbing on the host - the setup process will remain unchanged.
+2. Report the link state of SR-IOV NICs.
```
Report or management as well?

orelmisan replied:
Both.

xpivarc reviewed (commented):
I think that I have missed personally is description of how changes to the VM's interfaces are or not propagated to the VMI's interfaces.

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/39c035213a77fc3e34333de1ae0f76360c071365..26d6967dbd33764e5b0678b629d44ab80b0523de): Addressed @xpivarc's [comment](https://github.com/kubevirt/enhancements/pull/2#discussion_r1923727875).

maiqueb commented on veps/sig-network/link-state-proposal.md:
```diff
@@ -0,0 +1,230 @@
+- Start the VM with one or more interfaces with their link set to `down`.
+- Toggle the link state of one or more interfaces of a running VM.
+- Hot plug an interface with a link state set to `down`.
+- Control the link state of interfaces connected to primary and secondary networks.
```
By primary networks, do we actually desire to do this to the cluster-default-network ? 

I could maybe understand this for the primary UDN (ovn-k) feature, but for the cluster default network, that's maybe kind of breaking the kubernetes networking model ? Since the pod wouldn't be networked ... 

Do we even have use-cases for the cluster default network @phoracek ? 

Another thing that makes me itch a bit regarding the cluster default network is probing: assume I have a readiness / liveness prove. But I've set the interface link down because <*insert whatever weird use-case*>. This pod will crash-loop until the user enables the interface. It could actually turn out this is useful though ... Right now, I fail to see how. Maybe this could actually be useful for chaos testing 0_o

ormergi replied:
>I could maybe understand this for the primary UDN (ovn-k) feature, but for the cluster default network, that's maybe kind of breaking the kubernetes networking model ? Since the pod wouldn't be networked ...

I think its fine if we changes the interface state of the guest (using virsh).
With that the pod will remain connected to the default cluster network.

maiqueb replied:
Ah, right ! I keep forgetting this affect the VM iface, not the pod. 

Sorry.

EdDev replied:
This has been raised [here](https://github.com/kubevirt/enhancements/pull/2#discussion_r1916719655).

> the primary UDN (ovn-k) feature

I think this is a D/S feature, not a Kubevirt one.

> Do we even have use-cases for the cluster default network @phoracek ?

Yes, see the ref above.

> Another thing that makes me itch a bit regarding the cluster default network is probing: assume I have a readiness / liveness prove. But I've set the interface link down because <insert whatever weird use-case>. This pod will crash-loop until the user enables the interface. It could actually turn out this is useful though ... Right now, I fail to see how. Maybe this could actually be useful for chaos testing 0_o

In the end, it is the operator or VM owner responsibility.
I agree it should be warned about in the documentation, but to take a decision for all users may be one step too far.
I agree it may cause some issues when used incorrectly, but the stakeholders think it may be useful nevertheless.

One advantage I see with having this functional for all interfaces is that it is consistent and predictable.

> > I could maybe understand this for the primary UDN (ovn-k) feature, but for the cluster default network, that's maybe kind of breaking the kubernetes networking model ? Since the pod wouldn't be networked ...

> I think its fine if we changes the interface state of the guest (using virsh).
> With that the pod will remain connected to the default cluster network.

With bridge based bindings it will break for sure.

maiqueb replied:
It would still affect the probes that reach into the guest though.

IIRC that's not supported in OpenShift ... Would it made sense to open a new section about that ?

phoracek replied:
It would make sense that readiness probe fails if the primary iface is down - the VM obviously should not be a member of Service Endpoints. So the current behavior is fine IMO.

However, failing the liveness probe would force the VM to restart, which would make this feature unusable. We can try to be smart about this and ignore the network-based liveness probe on interfaces that were set down by the VM owner. @orelmisan could you tell how much extra effort / risk that would be?

We could say that the user opted out of this feature by using the probe. However, the use may not know in advance they will need this feature - e.g. that this particular VM will need to be fenced after it becomes noisy.

EdDev replied:
> IIRC that's not supported in OpenShift ... Would it made sense to open a new section about that ?

What d/s projects support or not are their business. d/s projects may choose to act/support a subset of of Kubevirt features.

> We can try to be smart about this and ignore the network-based liveness probe on interfaces that were set down by the VM owner. @orelmisan could you tell how much extra effort / risk that would be ... or if it's possible at all.

“Smart” is usually complex and sometime non-trivial.
I tend not to support such paths.

The limitation to block actions on the primary was raised early. If you prefer now to get back to it, we can.
Supporting it in follow ups can be discussed later, but with lower chance to get it into v1.5 release.

orelmisan replied:
I will create a VM with a primary network and probes, use virsh to set the link to `down`, and get back to you with the results.

I don't think we should complicate this simple feature just to play nice with another feature.
A limitation is a better path IMO.

orelmisan replied:
> I mean to say, I think the design should have a disclaimer / section indicating limitations and list guest side probes as one of them.

Done.

maiqueb reviewed (commented):
Just stirring the discussion around the primary network topic.

orelmisan commented:
[Change](https://github.com/kubevirt/enhancements/compare/26d6967dbd33764e5b0678b629d44ab80b0523de..7afe4796d1e87a6a57a243791e20bf533e45211b): Added a disclaimer about setting the primary interface's link state to down when using readiness / liveness probes.

maiqueb commented:
Tks for listening, and taking into account my suggestions. Re-adding the LGTM. 

/lgtm
/hold cancel
